src/psbt/mod.rs

```diff
@@ -861,6 +800,160 @@ impl PsbtExt for Psbt {
     }
 }
 
+/// Extension trait for PSBT inputs
+pub trait PsbtInputExt {
+    /// Given the decriptor for utxo being spent, populate the input with all the required fields to
+    /// spend it.
+    ///
+    /// If the descriptor contains wilcards or otherwise cannot be transformed into a concrete
+    /// descriptor an error will returned.
+    ///
+    /// You can make sure the descriptor is consistent with the `witness_utxo` and/or
+    /// `non_witness_utxo` fields of the PSBT by setting `check_witness_utxo` and
+    /// `check_non_witness_utxo` respectively. You should set `check_non_witness_utxo` to inputs the
+    /// output index in the transaction being spent.
+    fn update_with_descriptor(
+        &mut self,
+        descriptor: &Descriptor<DescriptorPublicKey>,
+        check_witness_utxo: bool,
+        check_non_witness_utxo: Option<u32>,
+    ) -> Result<(), UpdateInputError>;
+}
+
+impl PsbtInputExt for psbt::Input {
+    fn update_with_descriptor(
+        &mut self,
+        descriptor: &Descriptor<DescriptorPublicKey>,
+        check_witness_utxo: bool,
+        check_non_witness_utxo: Option<u32>,
+    ) -> Result<(), UpdateInputError> {
+        use std::cell::RefCell;
+        let mut inp = self.clone();
+        let secp = secp256k1::Secp256k1::verification_only();
+
+        let derived = if let Descriptor::Tr(_) = &descriptor {
+            // have to use a RefCell because we can't pass FnMut to translate_pk2
+            let tap_key_origins = RefCell::new(BTreeMap::new());
+            let derived = descriptor
+                .translate_pk2(|xpk| {
+                    let derived = xpk.derive_public_key(&secp)?;
+                    tap_key_origins.borrow_mut().insert(
+                        derived.to_x_only_pubkey(),
+                        (
+                            vec![/* we'll populate this in the next loop */],
+                            (xpk.master_fingerprint(), xpk.full_derivation_path()),
+                        ),
+                    );
+                    Ok(derived)
+                })
+                .map_err(UpdateInputError::DerivationError)?;
+            let mut tap_key_origins = tap_key_origins.into_inner();
+
+            // NOTE: the derived descriptor will always be Tr
+            if let Descriptor::Tr(tr) = &derived {
+                let spend_info = tr.spend_info();
+                for (_depth, ms) in tr.iter_scripts() {
+                    let leaf_script = (ms.encode(), LeafVersion::TapScript);
+                    let tapleaf_hash = TapLeafHash::from_script(&leaf_script.0, leaf_script.1);
+                    let control_block = spend_info
+                        .control_block(&leaf_script)
+                        .expect("Control block must exist in script map for every known leaf");
+                    inp.tap_scripts.insert(control_block, leaf_script);
+
+                    for pk in ms.iter_pk() {
```

Comment:
This should either be `iter_pk_pkh` or `translate_pk2`. When we iterate `iter_pk`, we are not iterating the `pkh` fragments. 

Aside: Maybe I will PR later to rename `iter_pk` to `iter_only_pk` to highlight this

Reply:
Hmm ok this will require me to index the data by the hash of the key rather than the xonly key. Is it really possible to have a pubkey hash of an xonly key in a tapscript :weary:?

Reply:
The Hash is the same type as of the Key. For `DescriptorPublicKey`, `Pk == Pk::Hash`

Reply:
This is `bitcoin::Publickey` here though. We're working with the the concrete one because we need it to compute the tapleaf hash with the script pubkey. My clunky plan would be to:

1. When deriving `DecriptorPublicKey` create an index from pubkey hash to `(xonly, fingerprint,  derivation path)`
2. When iterating over pubkey hashes in each left read the index and insert what I find in the index into `tap_leaf_origins` (if it hasn't been inserted yet) and add the leaf.

Quite awkward but I'm not sure there's another good way to do.

Reply:
I think you can have both things together in the for loop like

```rust
for ((_, ms_derived), (_, ms_underived) in tr.iter_scripts().zip(underived_tr.iter_scripts)
```

Reply:
Iterator is guaranteed to be consistent to traverse the leaves in the same order

Reply:
ah ok sure. I was thinking about that but wasn't sure I wanted to go rely on them iterating in the same order.. but I guess as long as we don't use `std` hashmaps for anything it should be fine!

Reply:
Yeah, they are guaranteed to output leaves in the dfs walk order.